### PR TITLE
Add support for different arguments per workspace

### LINKED ISSUE #663: Add support for different arguments per workspace
I am adding support for workspace folder specific settings in the vscode extension. I would like to be able to set things like `--build-path` and `-Xswiftc` arguments etc based on which workspace the file I am editing is in. At the moment we can only set these globally for all workspaces via sourcekit-lsp command line arguments.

### PR DESCRIPTION
**Pull request by ahoppen**
Fixes #663
rdar://101815704

@adam-fowler Do you think these new options on `WorkspaceFolder` would work for you? You should be able to specify them on both the `initialize` and the `workspace/didChangeWorkspaceFolders` request.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by adam-fowler** [comment]
I think these should cover most things. It would be good to also include support for SDKs when support for these is added.

**Comment by ahoppen** [comment]
> It would be good to also include support for SDKs when support for these is added.

I agree. Once we add that the SDK option to the command line options of `sourcekit-lsp`, we should also add it to the `WorkspaceFolder` options.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

File: Sources/SourceKitLSP/SourceKitServer.swift [hunk 1]
```diff
@@ -963,24 +964,57 @@ extension SourceKitServer: BuildSystemDelegate {
...
+  private func buildSetup(for workspaceFolder: WorkspaceFolder) -> BuildSetup {
+    let buildParams = workspaceFolder.buildSetup
+    let scratchPath: AbsolutePath?
+    if let scratchPathParam = buildParams?.scratchPath {
+      scratchPath = try? AbsolutePath(validating: scratchPathParam.pseudoPath)
+    } else {
+      scratchPath = nil
+    }
+    return SKCore.BuildSetup(
+      configuration: workspaceFolder.buildSetup?.buildConfiguration?.configuration,
```
Comment (hamishknight) [inline Sources/SourceKitLSP/SourceKitServer.swift:993]:
```suggestion
      configuration: buildParams?.buildConfiguration?.configuration,
```

File: Sources/SourceKitLSP/SourceKitServer.swift [hunk 2]
```diff
@@ -963,24 +964,57 @@ extension SourceKitServer: BuildSystemDelegate {
...
     }
+    let workspaceBuildSetup = self.buildSetup(for: workspaceFolder)
     return try? await Workspace(
       documentManager: self.documentManager,
-      rootUri: uri,
+      rootUri: workspaceFolder.uri,
       capabilityRegistry: capabilityRegistry,
       toolchainRegistry: self.toolchainRegistry,
-      buildSetup: self.options.buildSetup,
+      buildSetup: workspaceBuildSetup.merging(self.options.buildSetup),
```
Comment (hamishknight) [inline Sources/SourceKitLSP/SourceKitServer.swift:1017]: Is it expected that the main options can override the workspace options? I would have expected it to be the other way around

**Review by hamishknight — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

File: Sources/SourceKitLSP/SourceKitServer.swift (hunk 2 above)
Comment (ahoppen) [reply Sources/SourceKitLSP/SourceKitServer.swift:1017]: OK, I’ll flip the precedence order.

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Pull request merged**